examples/step-77/step-77.cc

```diff
@@ -0,0 +1,655 @@
+  // but the following function will be easy to understand if you understood
+  // what `assemble_system()` in step-15 did. Importantly, however, we need to
+  // compute the residual not linearized around the current solution vector, but
+  // whatever we get from KINSOL. This is necessary for operations such as line
+  // search where we want to know what the residual $F(U^k + \alpha_k \delta
+  // U^K)$ is for different values of $\alpha_k$; KINSOL in those cases simply
+  // gives us the argument to the function $F$ and we then compute the residual
+  // $F(\cdot)$ at this point.
+  //
+  // The function prints the norm of the so-computed residual at the end as a
+  // way for us to follow along the progress of the program.
+  template <int dim>
+  void MinimalSurfaceProblem<dim>::compute_residual(
+    const Vector<double> &evaluation_point,
+    Vector<double> &      residual)
+  {
+    TimerOutput::Scope t(computing_timer, "assembling the residual");
+
+    std::cout << "  Computing residual vector..." << std::flush;
+
+    const QGauss<dim> quadrature_formula(fe.degree + 1);
+    FEValues<dim>     fe_values(fe,
+                            quadrature_formula,
+                            update_gradients | update_quadrature_points |
+                              update_JxW_values);
+
+    const unsigned int dofs_per_cell = fe.n_dofs_per_cell();
+    const unsigned int n_q_points    = quadrature_formula.size();
+
+    Vector<double>              cell_residual(dofs_per_cell);
+    std::vector<Tensor<1, dim>> evaluation_point_gradients(n_q_points);
+
+    std::vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
+
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      {
+        cell_residual = 0;
+        fe_values.reinit(cell);
+
+        fe_values.get_function_gradients(evaluation_point,
+                                         evaluation_point_gradients);
+
+
+        for (unsigned int q = 0; q < n_q_points; ++q)
+          {
+            const double coeff =
+              1. / std::sqrt(1 + evaluation_point_gradients[q] *
+                                   evaluation_point_gradients[q]);
+
+            for (unsigned int i = 0; i < dofs_per_cell; ++i)
+              cell_residual(i) = (fe_values.shape_grad(i, q) // \nabla \phi_i
+                                  * coeff                    // * a_n
+                                  * evaluation_point_gradients[q] // * u_n
+                                  * fe_values.JxW(q));            // * dx
+          }
+
+        cell->get_dof_indices(local_dof_indices);
+        for (unsigned int i = 0; i < dofs_per_cell; ++i)
+          residual(local_dof_indices[i]) += cell_residual(i);
+      }
+
+    hanging_node_constraints.condense(residual);
+
+    for (types::global_dof_index i :
+         DoFTools::extract_boundary_dofs(dof_handler))
+      residual(i) = 0;
+
+    // TODO: think about whether this is actually necessary, and whether step-15
```

Comment:
:)

Reply:
We do similar in step-44 as well -- but there its done for all (Dirichlet) constrained DoFs. I don't think that one can meaningfully associate a residual with these DoFs, because their assigned values are controlled by the constraints themselves. So, IMHO, it might be even more correct to zero all of the entries constrained in the call to `MatrixTools::apply_boundary_values()` as well.